RESOURCE_PROVIDER_PLUGINS.md

```diff
@@ -0,0 +1,340 @@
+  - Resource id
+  - Submitted resource data.
+
+Upon successful completion of these operations the request will then be passed to the registered resource provider plugin.
+
+---
+## Resource Provider plugin:
+
+For a plugin to be considered a Resource Provider it needs to implement the `ResourceProvider` interface.
+
+By implementing this interface the plugin is able to register with the SignalK server the:
+- Resource types provided for by the plugin
+- Methods to used to action requests. 
+
+It is these methods that perform the retrival, saving and deletion of resources from storage.
+
+
+### Resource Provider Interface
+
+---
+The `ResourceProvider` interface defines the contract between the the Resource Provider plugin and the SignalK server and has the following definition _(which it and other related types can be imported from `@signalk/server-api`)_:
+
+```typescript
+interface ResourceProvider: {
+  types: string[],
+  methods: {
+    listResources: (type:string, query: {[key:string]:any})=> Promise<any>
+    getResource: (type:string, id:string)=> Promise<any>
+    setResource: (type:string, id:string, value:{[key:string]:any})=> Promise<any>
+    deleteResource: (type:string, id:string)=> Promise<any>
+  }
+}
+```
+where:
+
+- `types`: An array containing a list of resource types provided for by the plugin. These can be a mixture of both __Common__ and __Custom__ resource types.
+- `methods`: An object containing the methods resource requests are passed to by the SignalK server. The plugin __MUST__ implement each method, even if that operation is not supported by the plugin!
+
+#### __Method Details:__
+
+---
+__`listResources(type, query)`__: This method is called when a request is made for resource entries of a specific resource type that match a specifiec criteria.
+
+_Note: It is the responsibility of the resource provider plugin to filter the resources returned as per the supplied query parameters._
+
+`type:` String containing the type of resource to retrieve.
+
+`query:` Object contining `key | value` pairs repesenting the parameters by which to filter the returned entries. _e.g. {distance,'50000}_
+
+`returns:` 
+- Resolved Promise containing a list of resource entries on completion. 
+- Rejected Promise containing an Error if incomplete or not implemented.
+
+
+_Example resource request:_ 
+```
+GET /signalk/v1/api/resources/waypoints?bbox=5.4,25.7,6.9,31.2&distance=30000
+```
+_ResourceProvider method invocation:_
+
+```javascript
+listResources(
+  'waypoints', 
+  {
+    bbox: '5.4,25.7,6.9,31.2',
+    distance: 30000
+  }
+);
+```
+
+---
+__`getResource(type, id)`__: This method is called when a request is made for a specific resource entry of the supplied resource type and id.
+
+`type:` String containing the type of resource to retrieve.
+
+`id:` String containing the target resource entry id. _e.g. 'urn:mrn:signalk:uuid:07894aba-f151-4099-aa4f-5e5773734b99'_
```

Comment:
True... since the preamble has to be stripped off before storing anyaway as `:` is not a valid character for filenames when persisting to the filesystem.
I will update to exclude the preamble and modify the code to strip it before passing it to the plugin.

Reply:
I thought that the spec specified just uuid format, but it does include the full thing https://github.com/SignalK/specification/blob/master/schemas/groups/resources.json#L108 so on API level we should then stick to that...